### ISSUE TITLE
Support BuildBox for local builds

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#719]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/719)
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Oct 19, 2018, 15:00

BuildBox implements a FUSE filesystem to provide access to CAS contents without the staging overhead or danger of hardlinks.
It will also abstract sandboxing across platforms. BuildBox is currently used in BuildGrid workers for remote execution.

Switching to BuildBox also for local builds will thus reduce build (staging) time and allows us to remove platform-specific sandboxing code from BuildStream.

# Tasks
* [x] Completely optional support for developers/testing, see !951 
* [x] Support buildbox-run as sandboxing backend: #1177
* [x] Support workspace builds, depends on #985
* [x] Move sandboxing code from buildbox-fuse to buildbox-run-bubblewrap: https://gitlab.com/BuildGrid/buildbox/buildbox-fuse/issues/33. This depends on https://gitlab.com/BuildGrid/buildbox/buildbox-casd/issues/8
* [x] Support read-only rootfs, see https://gitlab.com/BuildStream/buildbox/issues/23 (also affects remote execution)
* [x] Configurable sandbox UID/GID, depends on https://gitlab.com/BuildStream/buildbox/issues/26
* [x] Consider potential issues due to CAS not storing timestamps (also affects remote execution)
* [x] Check whether command exists in virtual root directory
* [x] Support interactive mode for bst shell, depends on https://gitlab.com/BuildStream/buildbox/issues/27. Verify that job control works properly.
* [x] Support network access with bst shell, depends on https://gitlab.com/BuildStream/buildbox/issues/25
* [x] Update documentation to account for BuildBox as dependency
* [x] Add BuildBox and dependencies to CI images
* [x] Require virtual directory support for element plugins: #1262
* [x] Optimize output_directories in SandboxREAPI: #1223
* [x] Check for performance regressions
* [x] Fix any remaining test failures (currently marked as xfail)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Nov 13, 2018, 16:58

Mark as a blocker as I think this was one of the features we want for 1.4

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Nov 13, 2018, 17:25

FYI this wasn't discussed as a blocker during the gathering, or mentioned in the mail outlining targets for 1.4

https://mail.gnome.org/archives/buildstream-list/2018-October/msg00080.html

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Nov 13, 2018, 18:36

mentioned in merge request !920

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Nov 13, 2018, 18:36

Hey, thanks for that link; it was mentioned in the GoogleDoc we created together as a "nice to have".

If we know this is not going to happen then maybe we should reconsider the approach for https://gitlab.com/BuildStream/buildstream/issues/759 explained at https://gitlab.com/BuildStream/buildstream/issues/759#note_116418552

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Nov 14, 2018, 10:28

mentioned in merge request !951

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Apr 29, 2019, 10:00

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @willsalmon]](https://gitlab.com/willsalmon) on May 15, 2019, 11:51

assigned to [[Gitlab user @willsalmon]](https://gitlab.com/willsalmon)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @willsalmon]](https://gitlab.com/willsalmon) on Jun 26, 2019, 15:49

mentioned in merge request !1429

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @traveltissues]](https://gitlab.com/traveltissues) on Jul 12, 2019, 11:18

assigned to [[Gitlab user @traveltissues]](https://gitlab.com/traveltissues)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @traveltissues]](https://gitlab.com/traveltissues) on Jul 15, 2019, 13:26

[[Gitlab user @willsalmon]](https://gitlab.com/willsalmon) are the task statuses on this issue up-to-date?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @willsalmon]](https://gitlab.com/willsalmon) on Jul 15, 2019, 13:31

yes but !951 will complete the 1st, 7th, 11th and 12th and make progress towards 14th.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @traveltissues]](https://gitlab.com/traveltissues) on Jul 22, 2019, 16:58

marked the task **Configurable sandbox UID/GID, depends on https://gitlab.com/BuildStream/buildbox/issues/26** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @traveltissues]](https://gitlab.com/traveltissues) on Jul 22, 2019, 17:03

unassigned [[Gitlab user @traveltissues]](https://gitlab.com/traveltissues)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @willsalmon]](https://gitlab.com/willsalmon) on Jul 26, 2019, 10:41

marked the task **Completely optional support for developers/testing, see !951** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @willsalmon]](https://gitlab.com/willsalmon) on Jul 26, 2019, 10:41

marked the task **Check whether command exists in virtual root directory** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @willsalmon]](https://gitlab.com/willsalmon) on Jul 26, 2019, 10:42

marked the task **Fall back to bubblewrap/chroot backend for element plugins that don't support the virtual directory API (as is already in place for remote execution), i.e., replace the hacks in _platform/linux.py and sandbox/sandbox.py** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @willsalmon]](https://gitlab.com/willsalmon) on Jul 26, 2019, 10:45

I have updated some of the Todo items now that https://gitlab.com/BuildStream/buildstream/merge_requests/951 has landed but while this dose have a CI stage it has some issues please see https://gitlab.com/BuildStream/buildstream/commit/b54c8cb07ca257be79940ffa70853bf75d2c287c a commit undoing this commit and fixing the tests that are failing is needed before the CI TODO item on this issue can be completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Aug 8, 2019, 12:54

> while this dose have a CI stage it has some issues please see https://gitlab.com/BuildStream/buildstream/commit/b54c8cb07ca257be79940ffa70853bf75d2c287c a commit undoing this commit and fixing the tests that are failing is needed before the CI TODO item on this issue can be completed

This has been resolved in !1525.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Aug 8, 2019, 12:54

marked the task **Add BuildBox and dependencies to CI images** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Aug 8, 2019, 12:55

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Aug 8, 2019, 12:58

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Aug 8, 2019, 13:03

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Aug 8, 2019, 13:11

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Aug 8, 2019, 13:34

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Aug 8, 2019, 13:45

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Oct 21, 2019, 13:04

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tlater]](https://gitlab.com/tlater) on Nov 21, 2019, 17:48

[[Gitlab user @juergbi]](https://gitlab.com/juergbi) Can this be ticked off?

> * [ ] Make it usable on systems without user namespaces, depends on https://gitlab.com/BuildStream/buildbox/issues/26

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @danielsilverstone-ct]](https://gitlab.com/danielsilverstone-ct) on Nov 25, 2019, 11:07

assigned to [[Gitlab user @juergbi]](https://gitlab.com/juergbi) and unassigned [[Gitlab user @willsalmon]](https://gitlab.com/willsalmon)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Nov 28, 2019, 06:21

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Dec 10, 2019, 16:58

marked the task **Support buildbox-run as sandboxing backend: #1177** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Dec 10, 2019, 16:59

marked the task **Move sandboxing code from buildbox-fuse to buildbox-run-bubblewrap: https://gitlab.com/BuildGrid/buildbox/buildbox-fuse/issues/33. This depends on https://gitlab.com/BuildGrid/buildbox/buildbox-casd/issues/8** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) on Dec 16, 2019, 09:39

> * [ ] Support interactive mode for bst shell, depends on https://gitlab.com/BuildStream/buildbox/issues/27. Verify that job control works properly.
> * [ ] Support network access with bst shell, depends on https://gitlab.com/BuildStream/buildbox/issues/25

Should that (also?) be supported by buildbox-common as part of the runner interface? If so, do we need it on buildbox-fuse at all?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Dec 16, 2019, 09:49

These two should indeed be (optional) extensions to the runner interface and are not related to the now separate FUSE part. I.e., we should move these issues to buildbox-common and/or buildbox-run-bubblewrap.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Dec 16, 2019, 09:52

No, only the UID/GID part was implemented (and that will require additional work now that we use the buildbox-run interface).

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Jan 21, 2020, 09:28

mentioned in merge request !1807

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 11, 2020, 16:09

marked the task **Optimize output_directories in SandboxREAPI: #1223** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 11, 2020, 18:03

marked the task **Configurable sandbox UID/GID, depends on https://gitlab.com/BuildStream/buildbox/issues/26** as incomplete

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 12, 2020, 15:57

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 20, 2020, 10:57

marked the task **Support workspace builds, depends on #985** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 20, 2020, 10:58

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 20, 2020, 17:28

mentioned in merge request BuildGrid/buildbox/buildbox-run-bubblewrap!8

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 20, 2020, 17:30

mentioned in merge request !1823

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 25, 2020, 16:53

mentioned in merge request BuildGrid/buildbox/buildbox-run-bubblewrap!9

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 26, 2020, 09:14

mentioned in merge request !1825

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 26, 2020, 16:19

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 27, 2020, 13:07

marked the task **Configurable sandbox UID/GID, depends on https://gitlab.com/BuildStream/buildbox/issues/26** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 27, 2020, 13:07

marked the task **Support network access with bst shell, depends on https://gitlab.com/BuildStream/buildbox/issues/25** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 27, 2020, 13:07

marked the task **Consider potential issues due to CAS not storing timestamps (also affects remote execution)** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 27, 2020, 14:39

marked the task **Support read-only rootfs, see https://gitlab.com/BuildStream/buildbox/issues/23 (also affects remote execution)** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Mar 2, 2020, 18:03

mentioned in merge request !1826

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Mar 3, 2020, 09:44

marked the task **Support interactive mode for bst shell, depends on https://gitlab.com/BuildStream/buildbox/issues/27. Verify that job control works properly.** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Mar 4, 2020, 05:45

mentioned in merge request !1827

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Apr 23, 2020, 17:12

marked the task **Check for performance regressions** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Apr 23, 2020, 17:12

marked the task **Fix any remaining test failures (currently marked as xfail)** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Apr 27, 2020, 12:35

mentioned in merge request !1887

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Apr 28, 2020, 07:49

marked the task **Require virtual directory support for element plugins: #1262** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Apr 28, 2020, 07:57

mentioned in merge request !1888

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Apr 28, 2020, 12:16

marked the task **Update documentation to account for BuildBox as dependency** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @marge-bot123]](https://gitlab.com/marge-bot123) on Apr 29, 2020, 08:02

closed via merge request !1888

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @marge-bot123]](https://gitlab.com/marge-bot123) on Apr 29, 2020, 08:02

mentioned in commit 82eb1d4271bb634f248bc9e1770119d2815d7cd6
